### PR TITLE
Support public API keys

### PR DESCRIPTION
**Pull request by hiroara**
@danimal141 @iusztin I added changes to support [autorization with public keys](https://api.scnnr.cubki.jp/v1/docs#tag/authorization%2Fpaths%2F~1auth~1tokens%2Fpost).
After merging this, this library will be able to receive `publicAPIKey` and use it before requests that need authorization.
About the specification of public API keys, please read `README.md` and the API docs.

Please review it!

### CODE REVIEW

File: src/authInterceptor/PublicKeyAuthInterceptor.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  getOneTimeToken() {
+    return this.retrieveOneTimeToken()
+      .then(() => {
+        const token = this.oneTimeToken
+        delete this.oneTimeToken
+        return token
+      })
+  }
+
+  retrieveOneTimeToken() {
```
Comment (iusztin) [inline src/authInterceptor/PublicKeyAuthInterceptor.js]: I wonder if the functions with the naming `verb`+`OneTimeToken()` are a good candidate for making a `OneTimeToken` class (with `OneTimeToken.get()`, etc).
Then the `Interceptor` would only handle logic about intercepting.

File: dist/scnnr.cjs.js [hunk 2]
```diff
@@ -198,11 +198,140 @@ var errors = Object.freeze({
...
+    _this.marginToExpire = 0.05; // a margin to prevent unexpected expiration (5% of the time)
+    return _this;
+  }
+
+  createClass(PublicKeyAuthInterceptor, [{
+    key: 'interceptRequest',
+    value: function interceptRequest(config) {
+      return this.getOneTimeToken().then(function (token) {
+        config.headers['x-api-key'] = 'use-scnnr-one-time-token';
+        config.headers['x-scnnr-one-time-token'] = token;
```
Comment (iusztin) [inline dist/scnnr.cjs.js]: Your implementation seems to work well! The following comment is maybe for a future PR on the scnnr project.

I think maybe our documentation needs to be updated here: https://api.scnnr.cubki.jp/v1/docs#tag/authorization%2Fpaths%2F~1auth~1tokens%2Fpost

Even though the header field is `x-scnnr-one-time-token` (which is used successfully in this implementation),
the documentation specifies it as `use-x-scnnr-one-time-token` and as `-H 'use-scnnr-one-time-token: ..`

File: test/Connection.js [hunk 3]
```diff
@@ -20,14 +20,36 @@ describe('Connection', () => {
...
   }
-  const connection = new scnnr.Connection(config)
   const responseBody = { data: 'dummy_data' }
 
+  function getConnection(cfg = config) { return new scnnr.Connection(cfg) }
+
+  describe('constructor', () => {
+    const requestPath = '/recognitions/some/recognition-id'
+
+    context('when an authInterceptor is pass', () => {
```
[the remark this answers is not in the document]
Comment (hiroara) [reply test/Connection.js]: Ah, exactly! I'll fix it.

File: test/Connection.js [hunk 4]
```diff
@@ -20,14 +20,36 @@ describe('Connection', () => {
...
+
+      it('registers it for requests', () => {
+        const connection = getConnection(Object.assign({ authInterceptor }, config))
+        expect(connection.httpClient.interceptors.request.handlers.length).to.equal(1)
+        expect(connection.httpClient.interceptors.request.handlers[0].fulfilled)
+          .to.equal(authInterceptor.interceptRequest)
+      })
+    })
+
+    context('when an authInterceptor does not pass', () => {
```
[the remark this answers is not in the document]
Comment (hiroara) [reply test/Connection.js]: You're right. I'll fix it!

File: test/Connection.js [hunk 5]
```diff
@@ -20,14 +20,36 @@ describe('Connection', () => {
...
+      it('does not register any intercepter for requests', () => {
```
[the remark this answers is not in the document]
Comment (hiroara) [reply test/Connection.js]: Thank you for the good find!

File: src/authInterceptor/PublicKeyAuthInterceptor.js (hunk 1 above)
Comment (hiroara) [reply src/authInterceptor/PublicKeyAuthInterceptor.js]: Good point! I'll try to consider refactoring.

File: dist/scnnr.cjs.js (hunk 2 above)
Comment (hiroara) [reply dist/scnnr.cjs.js]: Exactly. That is the problem of our API server project.
And I already fixed it, but we don't deploy it yet.
I'll plan to deploy it as soon as possible, so please be patient. 😅

**Comment by hiroara** [comment]
@iusztin I pushed the changes. Please take a look 👍

File: src/OneTimeTokenProvider.js [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+  }
+
+  issue() {
+    return Connection.build(true, Object.assign({}, this.options, { apiKey: this.publicAPIKey }))
+      .sendJson('/auth/tokens', { type: 'one-time' })
+      .then(response => response.data)
+  }
+
+  storeToken(data) {
+    this.timeout = setTimeout(() => { this.token = null }, data.expires_in * (1 - this.marginToExpire) * 1000)
```
Comment (iusztin) [inline src/OneTimeTokenProvider.js]: I think a call to `clearTimeout(this.timeout)` is missing here.

Comment (hiroara) [reply src/OneTimeTokenProvider.js]: Ah, I totally forgot about it! Though I stored the return value from `setTimeout`...
I added the change. 👍

File: src/OneTimeTokenProvider.js [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+  }
+
+  get() { return this.retrieve().then(() => this.getAndClearToken()) }
+
+  retrieve() {
+    if (this.token != null) { return Promise.resolve() }
+    return this.issue().then(data => this.storeToken(data))
+  }
+
+  issue() {
```
Comment (iusztin) [inline src/OneTimeTokenProvider.js]: I had an idea to make the naming more distinct (compared to retrieve and get). How about renaming `issue()` to `requestToken()`, as it generates request to the server (to get a token)? And I think the server issues the token.. not the client.
What are your thoughts on this?

File: src/OneTimeTokenProvider.js [hunk 8]
```diff
@@ -0,0 +1,43 @@
+import Connection from './Connection'
+
+export default class OneTimeTokenProvider {
+  constructor(publicAPIKey, options) {
+    this.publicAPIKey = publicAPIKey
+    this.options = options
+    this.token = null
```
Comment (iusztin) [inline src/OneTimeTokenProvider.js]: I wonder if `this.token` and `this.timeout` would fit well into a `OneTimeToken` class. `OneTimeToken` could handle its validity and timeout by itself. And `OneTimeTokenProvider` would only take care of providing a valid `OneTimeToken`.

Comment (hiroara) [reply src/OneTimeTokenProvider.js]: Thank you for the suggestion 😃
I like such as the idea of using a value object.

But it's hard to imagine how we should deal with the cache and its expiration for me a little bit.
For example, what should own the cache? If it is `OneTimeTokenProvider`, and `OneTimeToken` should expire itself, how do the token clear the cache? In this case, I think `OneTimeToken` is owned by `OneTimeTokenProvider`, so `OneTimeToken` shouldn't have a reference to the provider.

Could you share more details?

File: src/OneTimeTokenProvider.js (hunk 7 above)
Comment (hiroara) [reply src/OneTimeTokenProvider.js]: I think issuing tokens by this class is not wrong.
According to OOP, the implementation details should be hidden. When we follow this idea, a caller should not know the implementation details.
In this case, `OneTimeTokenProvider` is the class that issues one-time tokens in an unknown way.
Maybe, this can be said as a delegation.
This is the reason why I think we can use the name `issue` for this method.
What do you feel about this thought?

File: src/OneTimeTokenProvider.js [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+    this.timeout = null
+    this.marginToExpire = 0.05 // a margin to prevent unexpected expiration (5% of the time)
+  }
+
+  get() { return this.retrieve().then(() => this.getAndClearToken()) }
+
+  retrieve() {
```
[the remark this answers is not in the document]
Comment (hiroara) [reply src/OneTimeTokenProvider.js]: Thank you for your idea!
But considering [this comment I wrote](https://github.com/NEWROPE/scnnr-js/pull/13/files/c00230983f99354eb29d1d30cd9d8b4f0bf1690e#r202901321), I guess we can also reorganize methods as follows:

- `get` => `get` (no change)
- `retrieve` => `issue` (issue a token, and if a cache exists, reuse it)
- `issue` => `requestToken` (send a request to issue a token by calling the API anyway)

What do you think?

File: src/OneTimeTokenProvider.js (hunk 7 above)
Comment (iusztin) [reply src/OneTimeTokenProvider.js]: Your point is very good. Implementation details should be hidden. And from the perspective of an unknowing caller, it looks like `OneTimeTokenProvider` issues a new token. I wasn't sure that this function will be used by a caller directly (to get a token without timeout or cache handling).

File: src/OneTimeTokenProvider.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (iusztin) [reply src/OneTimeTokenProvider.js]: I feel like it would be an improvement, as it reduces the amount of verbs with similar meanings, and the naming is a bit closer to the implementation.

File: src/OneTimeTokenProvider.js (hunk 8 above)
Comment (iusztin) [reply src/OneTimeTokenProvider.js]: I think in such a case `OneTimeToken` could have a flag `hasExpired()` if it has expired or not (-> the token handles expiration). And it could have a method to cause immediate expiration `expireNow()` (which would stop the timers and set the expiration flag).

And how about: `OneTimeTokenProvider` owns a `OneTimeToken`. Therefore `OneTimeTokenProvider` is responsible for handling/invalidating the cache/storing a token.

We could check if `someToken.hasExpired()` is true, to invalidate the cache and create a new token object. And when someone uses `oneTimeTokenProvider.get()` we could call `someToken.expireNow()` to expire the token immediately because it was used.

Hmm.. though now I'm not sure. It's not easy.. I'm happy to hear your thoughts.

Comment (hiroara) [reply src/OneTimeTokenProvider.js]: Did you mean a current-time-based logic instead of `setTimeout` by your suggestion?
In the current architecture, there is no special reason to pack a token value and expiration information, because `expires_in` is used only from `OneTimeTokenProvider`.
Considering to remove `setTimeout`, it would be beneficial because the timing to execute the callback is unstable, and complexity of asynchronous programming.
In conclusion, I think both would work, but to simplify, we can refactor by above reason.
How about this aspect?

Comment (iusztin) [reply src/OneTimeTokenProvider.js]: Considering the use of the token, I think you are right.

Ah, I didn't intend to suggest to remove `setTimeout`. But now that you said it, I think it's a good idea. I agree with your reasons.

Comment (hiroara) [reply src/OneTimeTokenProvider.js]: OK, thank you for the discussion! I'll apply the changes 👍

**Comment by hiroara** [comment]
@iusztin I added changes as we discussed. Please take a look!
